### DISCUSSION TITLE
Exploring new optimī optimizers

### DISCUSSION THREAD

**Discussion by mhirki** [bghira/SimpleTuner#819] [Show and tell]
I've been playing around with the new optimī optimizers. One of the new optimizers is Adan which is supposed to be superior to AdamW while using more VRAM:

https://optimi.benjaminwarner.dev/optimizers/adan/

The Adan optimizer requires a higher learning rate. I ended up setting it to 1e-3 which is 2.5 times higher than with AdamW. As expected, Adan is clearly converging faster. Given how slow Flux training is, Adan seems to be worth considering.

| Steps | adamw_bf16 | optimi-adan |
| --- | --- | --- |
| 500 | ![step_500_validation_('1024', '1024')](https://github.com/user-attachments/assets/1a528285-66eb-41ff-b827-2eaf674f9af2) | ![step_500_validation_('1024', '1024')](https://github.com/user-attachments/assets/031cc0f1-cda3-4e0f-a2b4-60ad60b00a6b) |
| 1000 | ![step_1000_validation_('1024', '1024')](https://github.com/user-attachments/assets/2e170f49-786d-46d3-9b45-335de292e391) | ![step_1000_validation_('1024', '1024')](https://github.com/user-attachments/assets/b9823eb1-e498-4c4f-9626-dff666e485d3) |
| 1500 | ![step_1500_validation_('1024', '1024')](https://github.com/user-attachments/assets/a1fb85df-dc91-4bca-bc95-5f870e87abcd) | ![step_1500_validation_('1024', '1024')](https://github.com/user-attachments/assets/cae9f881-6c88-4e1f-8b14-e78d7799518c) |
| 2000 | ![step_2000_validation_('1024', '1024')](https://github.com/user-attachments/assets/2dbcea87-4fad-4a80-b1d4-22e0bee32d9c) | ![step_2000_validation_('1024', '1024')](https://github.com/user-attachments/assets/843df507-6727-4911-9cc7-3f5d89f93eae) |
| 2500 | ![step_2500_validation_('1024', '1024')](https://github.com/user-attachments/assets/82e4fc73-8798-41cc-ab60-ea487a829bb3) | ![step_2500_validation_('1024', '1024')](https://github.com/user-attachments/assets/4c563852-b46a-4af7-a0c7-7747bd12c972) |
| 3000 | ![step_3000_validation_('1024', '1024')](https://github.com/user-attachments/assets/32325598-1f5d-4470-820e-419d433ca913) | ![step_3000_validation_('1024', '1024')](https://github.com/user-attachments/assets/6aa6cd02-b281-4dc6-be7f-c75ab47c6d35) |
| 3500 | ![step_3500_validation_('1024', '1024')](https://github.com/user-attachments/assets/597eaea6-cf09-42dd-a003-c6037ee17de5) | ![step_3500_validation_('1024', '1024')](https://github.com/user-attachments/assets/99eb06b3-c17e-4677-ae0a-6c4513415097) |
| 4000 | ![step_4000_validation_('1024', '1024')](https://github.com/user-attachments/assets/f90c5a90-f08f-4345-82af-06bc04337436) | ![step_4000_validation_('1024', '1024')](https://github.com/user-attachments/assets/661d15b4-412e-4987-bfbd-dd26e626e775) |
| 4500 | ![step_4500_validation_('1024', '1024')](https://github.com/user-attachments/assets/c27de462-807e-43f8-86b9-bba2ad077f4f) | ![step_4500_validation_('1024', '1024')](https://github.com/user-attachments/assets/a1611433-1816-440b-b2fc-18db1daf42c8) |
| 5000 | ![step_5000_validation_('1024', '1024')](https://github.com/user-attachments/assets/758ebfd9-d476-444f-9cff-bf65b750b57f) | ![step_5000_validation_('1024', '1024')](https://github.com/user-attachments/assets/dc863e0b-46a0-4bb5-b419-ada82d6ae605) |
| 5500 | ![step_5500_validation_('1024', '1024')](https://github.com/user-attachments/assets/0259a035-b071-43cf-98c4-77f75679b845) | ![step_5500_validation_('1024', '1024')](https://github.com/user-attachments/assets/02dfa718-a252-49d4-9b68-75ec5a691dde) |
| 6000 | ![step_6000_validation_('1024', '1024')](https://github.com/user-attachments/assets/533eb065-9d3a-45c2-aefa-d957e32cb506) | ![step_6000_validation_('1024', '1024')](https://github.com/user-attachments/assets/9fcbdaf6-5ac5-41ef-a656-f7c18cd18581) |

The LoRAs are available on the Huggingface Hub:
https://huggingface.co/mikaelh/flux-sanna-marin-v0.4-fp8-adamw-stochastic
https://huggingface.co/mikaelh/flux-sanna-marin-v0.4-fp8-adan

The exact config.env settings can also be found on the Hub.

**Comment by bghira** [comment]
oh cool! that's a pretty nice test. makes me think it could be useful for jamming more creativity into the model

**Comment by mhirki** [comment]
I added Lion to the mix:

https://huggingface.co/mikaelh/flux-sanna-marin-v0.4-fp8-lion

The challenge with Lion is that it converges extremely fast. I ended up dropping learning rate by 10x, increasing weight decay by 10x, increasing warmup steps to 400 and switching to polynomial schedule.

The final validation image at 6000 steps is looking good:
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![step_6000_validation_('1024', '1024')](https://github.com/user-attachments/assets/f3536e3b-25d1-416c-b47d-848dc3cbab0f) | ![step_6000_validation_('1024', '1024')](https://github.com/user-attachments/assets/0b31e66c-7d0e-429d-adc7-b2e11f6eeee0) | ![step_6000_validation_('1024', '1024')](https://github.com/user-attachments/assets/e24cb894-e012-47fe-b0a7-fc49ab14ebfe)

I did a little 3-way prompt comprehension test. AdamW seems to be the winner with Lion also performing well. Adan is the worst so maybe I need to try adjusting some parameters.

**sanna marin playing tennis**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724226077_0001](https://github.com/user-attachments/assets/33e2ef99-ea3b-4cd1-b759-71fc0cd9cc30) | ![flux_1724226159_0001](https://github.com/user-attachments/assets/b7da6d22-ad04-4086-94fc-44f9ee0ce60c) | ![flux_1724226118_0001](https://github.com/user-attachments/assets/be4ef13f-c985-44d0-94e9-eb83bb4e0440) |

**sanna marin running a marathron**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724226282_0001](https://github.com/user-attachments/assets/a5115758-7bca-4bef-bcd9-a2f7fd930eb2) | ![flux_1724226370_0001](https://github.com/user-attachments/assets/c740512f-596c-4743-8594-f1589aa05f48) | ![flux_1724226325_0001](https://github.com/user-attachments/assets/48c73903-2cca-4d85-a995-c59b46e712b2) |

**sanna marin sitting on a plane**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724226413_0001](https://github.com/user-attachments/assets/ff876ec3-6f54-48b9-85cd-5a4df7dc3892) | ![flux_1724226501_0001](https://github.com/user-attachments/assets/2fa7aee7-731e-4edc-a93f-163b45d04aa9) | ![flux_1724226457_0001](https://github.com/user-attachments/assets/ec9c9184-a5bb-4ce4-8214-20a76f0ca657) |

**sanna marin opening a can in the kitchen**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724226678_0001](https://github.com/user-attachments/assets/f9094f9c-303a-449e-a692-3b808aa7cfe9) | ![flux_1724226766_0001](https://github.com/user-attachments/assets/f130ea02-1e0d-4594-bb72-fd99b875ddc2) | ![flux_1724226722_0001](https://github.com/user-attachments/assets/fe375865-cf80-42f1-bd47-0a5a498c92de) |

**sanna marin petting a cat**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724226921_0001](https://github.com/user-attachments/assets/5c4aeb02-bc62-4bc3-8d64-ad4293413f43) | ![flux_1724226999_0001](https://github.com/user-attachments/assets/c83f0d70-c160-45ba-b885-dd1f423447f3) | ![flux_1724226960_0001](https://github.com/user-attachments/assets/83ccaea3-ad84-44e3-b7f9-4b3891f73ad7) |

**Comment by mhirki** [comment]
I managed to improve results with Adan by setting beta3 to 0.999 (instead of 0.99). Adan is still producing results which are different from AdamW and Lion with the same seed.

https://huggingface.co/mikaelh/flux-sanna-marin-v0.4-fp8-adan2

I added three new prompts to the comparison.

**sanna marin playing tennis**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724226077_0001](https://github.com/user-attachments/assets/fad4a974-b8eb-4608-abe3-1ef2462ca98f) | ![flux_1724276138_0001](https://github.com/user-attachments/assets/58ec3982-d7d3-4ab1-9097-01207765b03b) | ![flux_1724226118_0001](https://github.com/user-attachments/assets/9000ad75-ab35-4c02-868a-2768548d9968) |

**sanna marin running a marathon**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724226282_0001](https://github.com/user-attachments/assets/93716b2c-b3d3-46e1-b5f1-eb2d19c0840b) | ![flux_1724276169_0001](https://github.com/user-attachments/assets/3828d2be-44f3-4488-8632-e3058798da7c) | ![flux_1724226325_0001](https://github.com/user-attachments/assets/fc4db1f4-4e35-4f0c-b155-d0a4276994e1) |

**sanna marin sitting on a plane**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724226413_0001](https://github.com/user-attachments/assets/f0122323-8214-4582-acba-e80ae6f7ad05) | ![flux_1724276200_0001](https://github.com/user-attachments/assets/edef08ff-6f3c-4eb2-a3d6-bc368f00b39f) | ![flux_1724226457_0001](https://github.com/user-attachments/assets/a6d4c443-b6f2-4500-9613-8054559b9f6d) |

**sanna marin opening a can in the kitchen**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724226678_0001](https://github.com/user-attachments/assets/8b9007e0-1c7d-409f-9ce0-61af1c6e8a76) | ![flux_1724276232_0001](https://github.com/user-attachments/assets/c140fbf9-516c-4fb4-8ebd-c645d207dc88) | ![flux_1724226722_0001](https://github.com/user-attachments/assets/bd0028eb-c318-4c13-ac0e-a7a24d8c8f26) |

**sanna marin petting a cat**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724226921_0001](https://github.com/user-attachments/assets/f8118262-8baa-4c41-817f-7091d836b6a1) | ![flux_1724276263_0001](https://github.com/user-attachments/assets/d336d36c-cdcc-498e-8ab5-2d8d00821d71) | ![flux_1724226960_0001](https://github.com/user-attachments/assets/2907b8b6-f030-4874-a723-ae41045fef0b) |

**sanna marin catching a red ball**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724276678_0001](https://github.com/user-attachments/assets/763e66e3-6ab7-44db-b75e-d2dd0779a033) | ![flux_1724276636_0001](https://github.com/user-attachments/assets/b9df1849-5c46-489b-8bd4-2cb8f48d62ab) | ![flux_1724276719_0001](https://github.com/user-attachments/assets/9b29ec63-dd4b-4c1c-abec-6ac3a02ad677) |

**sanna marin picking up a strawberry**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724277613_0001](https://github.com/user-attachments/assets/a0223e80-2be3-42e9-a3ce-5b3af1d06561) | ![flux_1724277571_0001](https://github.com/user-attachments/assets/1a5107b0-9193-4c44-9f66-893a746d98f3) | ![flux_1724277656_0001](https://github.com/user-attachments/assets/9c85cd43-e205-43aa-b945-89e578bf8d86) |

**sanna marin picking up a strawberry**
| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![flux_1724315084_0001](https://github.com/user-attachments/assets/b1c7f5f4-3ed8-42bc-ab04-0c8082c8a6fa) | ![flux_1724315124_0001](https://github.com/user-attachments/assets/2ca862f3-7242-4226-8852-08a2c9f488c0) | ![flux_1724315162_0001](https://github.com/user-attachments/assets/4073cf34-94b1-43e5-b887-a7736d9c6b74) |

**Reply by mhirki** [reply]
Somewhat interestingly both Adan and Lion seem to have finished with a lower training loss than AdamW. AdamW's loss was very low around 3000 steps though.

| adamw_bf16 | optimi-adan | optimi-lion |
| --- | --- | --- |
| ![train_loss_adamw](https://github.com/user-attachments/assets/b9c858bb-aaea-43b8-b3b5-6f85df842d1d) | ![train_loss_adan3](https://github.com/user-attachments/assets/498d4fbd-a7b5-4b25-8e75-701a07f39fb9) | ![train_loss_lion4](https://github.com/user-attachments/assets/f0373677-59b9-4244-9389-9c14c8d3d4f9) |
